src/library/scala/concurrent/Future.scala

```diff
@@ -731,16 +732,24 @@ object Future {
     val i = futures.iterator
     if (!i.hasNext) Future.never
     else {
+      val deregs = ListBuffer.empty[() => Unit]
+      val l = new CountDownLatch(1)
       val p = Promise[T]()
       val firstCompleteHandler = new AtomicReference[Promise[T]](p) with (Try[T] => Unit) {
         override final def apply(v1: Try[T]): Unit =  {
           val r = getAndSet(null)
-          if (r ne null)
+          if (r ne null) {
             r tryComplete v1 // tryComplete is likely to be cheaper than complete
+            Future {
```

Comment:
Is spawning a Future here a good idea?

Reply:
I think it avoids the immediate deadlock, but it could in general cause a deadlock if futures like that exhaust the thread pool (see my comment below).